### ISSUE TITLE
Investigate Ingest Performance

### ISSUE THREAD

**Issue by hraberg** [xtdb/xtdb#328]
(card rewritten by Hugo)

One function with lots of parameters to start a custom ingestion performance test.

Things to measure:

- [x] Throughput

- [x] Possible slowdown on large enough ingestions

- [x] Latency (time until the data is available for query)

- [x] Disk space (distinguishing between document store and log)

- [x] Effect of sending the data in parallel to multiple nodes

- [x] Effect of sending a bunch of requests at once rather than sequentially (with Gatling?)

Should work with any combination of:

- [x] Kafka

- [x] JDBC with h2

- [x] JDBC with postgresql, mysql, sqlite

- [x] Oracle

and KV backend:

- [x] RocksDB

- [ ] LMDB

Things to ingest (in rough order of priority):

- [x] Watdiv (https://dsg.uwaterloo.ca/watdiv/)

- [ ] LUBM (http://swat.cse.lehigh.edu/projects/lubm/)

- [ ] IMDB

- [x] Timeseries (https://docs.timescale.com/v1.2/tutorials/other-sample-datasets#in-depth-devices https://docs.timescale.com/v1.2/tutorials/other-sample-datasets#in-depth-weather)

- [ ] Something bitemporal (we've yet to find a sample data set that would be easy to convert to EDN entities)

All of these datasets we've already worked on to some extent; see in `crux-test` or `docs/example`.

.

We can spike adding a thread pool / consumer group for the document topic, thought this is likely to complicate the already quite complex ingest logic, so it would have to carry its weight. But this would be assuming that serialisation and creation of the key values for the index is what's taking time. RocksDB has its own set of threads that take over once one hand the write to it. Tuning that can be looked at, but isn't really the intention of this card.

Also somewhat related to #270 as this also needs to sanity check the polling performance and most likely add indexes to make it work.

**Comment by hraberg** [comment]
One idea that has been raised before is to move the creation of the indexed key values to the client submitting the transaction and share the burden that way, but there are several issues with this, and it also makes some assumptions about where the bottleneck lies.

**Comment by hraberg** [comment]
Seems like a bit more than 10% of ingest time is spent updating the attribute statistics in a separate write.

**Comment by hraberg** [comment]
Some stats ingesting the lubm10.ntriples which is 100k triples.

First number is submit time, second sync time.
As can be seen (and expected), Kafka is the only one which results in asynchronous writes.

Kafka:
"Elapsed time: 2312.630106 msecs"
"Elapsed time: 5319.251471 msecs"

Moberg:
"Elapsed time: 92669.934058 msecs"
"Elapsed time: 1.665072 msecs"

JDBC (Postgres):
"Elapsed time: 212670.28064 msecs"
"Elapsed time: 0.2196 msecs"

JDBC (h2):
"Elapsed time: 593475.707901 msecs"
"Elapsed time: 0.248758 msecs"

JDBC (SQLite):
```
18:54:27.437 [crux.tx.event-log-consumer-thread] ERROR crux.tx.polling - Event log consumer threw exception, consumption has stopped:
java.sql.SQLException: [SQLITE_BUSY]  The database file is locked (database is locked)
	at org.sqlite.DB.newSQLException(DB.java:383)
	at org.sqlite.DB.newSQLException(DB.java:387)
	at org.sqlite.DB.execute(DB.java:339)
	at org.sqlite.PrepStmt.execute(PrepStmt.java:65)
	at next.jdbc.result_set$stmt__GT_result_set.invokeStatic(result_set.clj:416)
	at next.jdbc.result_set$stmt__GT_result_set.invoke(result_set.clj:411)
	at next.jdbc.result_set$eval21811$fn__21819.invoke(result_set.clj:470)
	at next.jdbc.protocols$eval21174$fn__21205$G__21165__21214.invoke(protocols.clj:33)
	at next.jdbc$execute_BANG_.invokeStatic(jdbc.clj:180)
	at next.jdbc$execute_BANG_.invoke(jdbc.clj:167)
	at crux.jdbc.JDBCEventLogConsumer$fn__22243.invoke(jdbc.clj:149)
	at next.jdbc.transaction$transact_STAR_.invokeStatic(transaction.clj:38)
	at next.jdbc.transaction$transact_STAR_.invoke(transaction.clj:18)
	at next.jdbc.transaction$eval21895$fn__21896.invoke(transaction.clj:78)
	at next.jdbc.protocols$eval21264$fn__21265$G__21255__21274.invoke(protocols.clj:57)
	at next.jdbc$transact.invokeStatic(jdbc.clj:205)
	at next.jdbc$transact.invoke(jdbc.clj:197)
	at crux.jdbc.JDBCEventLogConsumer.next_events(jdbc.clj:146)
	at crux.tx.polling$polling_consumer$fn__15926.invoke(polling.clj:31)
	at crux.tx.polling$polling_consumer.invokeStatic(polling.clj:18)
	at crux.tx.polling$polling_consumer.invoke(polling.clj:9)
	at crux.tx.polling$start_event_log_consumer$fn__15936.invoke(polling.clj:48)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:835)
```

**Comment by hraberg** [comment]
In JDBC, the call to `doc-exists?` seems to be a chunk of the time.

**Comment by hraberg** [comment]
In in Moberg, the slowness seems to mainly be stemming from the actual writing of events into RocksDB.

**Comment by hraberg** [comment]
Actually, I was wrong above, all systems will result in potentially waiting for sync, as the submission is decoupled from the indexing here's 

Moberg (MemKv):
"Elapsed time: 2891.235581 msecs"
"Elapsed time: 4059.248049 msecs"

As can be seen, this is similar to Kafka, so it's mainly the RocksDB writes that take time.

**Comment by hraberg** [comment]
The issue in Moberg is fsync related, if one doesn't provide a `:crux.standalone/event-log-sync-interval-ms` it will default to syncing every write to storage.

"Elapsed time: 2389.717786 msecs"
"Elapsed time: 4407.184518 msecs"

**Comment by hraberg** [comment]
With some tweaks, mainly reusing connections, and adding a flag and index for compaction `doc-exits?` check, numbers are:

JDBC (Postgres):
"Elapsed time: 4822.359689 msecs"
"Elapsed time: 3405.354368 msecs"

JDBC (h2):
"Elapsed time: 1966.702118 msecs"
"Elapsed time: 4510.916526 msecs"

Note that there's also a few missing `order by` statements when the log is queries, which results in random order when consuming the log. The sync time can be lowered by increasing the batch size, that might be the case for Moberg and Kafka as well.

**Comment by hraberg** [comment]
The flag is to avoid querying (and index) against blobs. In theory the flag can also remove the need for adding tombstones to the log.

**Comment by hraberg** [comment]
JDBC (SQLite):
"Elapsed time: 379563.646578 msecs"
"Elapsed time: 0.113429 msecs"

This is with a pool size of 1, as SQLite is meant to be used from a single thread, there are likely better solutions here.

**Comment by hraberg** [comment]
Here's an easy way to submit WatDiv, this uses the local dev node:
```
 (dev)
 (start)
 (set-log-level! 'crux.kafka :debug) ;; or 'crux.jdbc etc.
 (io/resource crux.watdiv-test/watdiv-triples-resource))]
       (rdf/submit-ntriples (:tx-log node) in 1000))
```

The data needs to be downloaded first, see top of the `crux.watdiv-test` namespace, the actual data is here: https://dsg.uwaterloo.ca/watdiv/watdiv.10M.tar.bz2

**Comment by hraberg** [comment]
You can inspect the result of the above load when it has finished with:
```
(crux/status node)
```
There's an (old) example in `hakan.clj` on how to run a subset of the queries (~250 or so), relevant parts:
```
  (swap! (:cache-state (:kv-store node)) empty) ;; clear all caches
  (defn run-watdiv-reference
    ([]
     (run-watdiv-reference (io/resource "watdiv/watdiv_crux.edn") #{35 67}))
    ([resource skip?]
     (let [times (mapv
                  (fn [{:keys [idx query crux-results]}]
                    (let [start (System/currentTimeMillis)]
                      (try
                        (let [result (count (.q (.db node)
                                                (crux.sparql/sparql->datalog query)))]
                          (assert (= crux-results result)
                                  (pr-str [idx crux-results result])))
                        (catch Throwable t
                          (prn idx t)))
                      (- (System/currentTimeMillis) start)))
                  (->> (read-string (slurp resource))
                       (remove :crux-error)
                       (remove (comp skip? :idx))))
           total (reduce + times)]
       {:total total :average (/ total (double (count times)))})))
(run-watdiv-reference)
```
Takes 30-90 seconds depending on various things.

**Comment by Roxolan0** [comment]
Edited OP with current goals.

**Comment by jarohen** [comment]
have split this out under #389
